elastic/elastic-agent
# [Upgrade Watcher] Try restarting Agent in multiple ways during rollback

## What does this PR do?

This PR fixes a potential bug where, during an upgrade, if the upgraded Agent process crashes, the Upgrade Watcher initiates a rollback but isn't able to restart the rolled back Agent.

Before this PR, the Upgrade Watcher would attempt to restart the rolled back Agent via a GRPC call to the Agent's `Restart()` function.  However, if the Agent process has crashed, there is no GRPC server available to receive such a call, thus preventing the Agent from being restarted. 

With this PR, the Upgrade Watcher continues to use the GRPC `Restart()` call to restart the Agent.  However, if that call fails, the Upgrade Watcher uses a fallback method of asking the Elastic Agent service to restart the Agent.

## Checklist



- [ ] ~I have added an integration test or an E2E test~ This bug was found while writing the integration test in https://github.com/elastic/elastic-agent/pull/3085.  As such, this bug fix will be tested by that test.

## How to test this PR locally

1. Build an Agent package from `main` (currently at version `8.11.0`), unpack it, and install it. This will serve as the version of Agent we will upgrade _from_.
   ```
   SNAPSHOT=true EXTERNAL=true DEV=true PLATFORMS=linux/arm64 PACKAGES=tar.gz mage package
   ```
3. Build an Agent package from the branch for this PR (say at version `8.12.0`). This will serve as the version of Agent we will upgrade _to_.  This Agent will deliberately crash upon start up but it also has the bugfix in the Upgrade Watcher code path.
   ```
   DEV=true PLATFORMS=linux/arm64 PACKAGES=tar.gz mage package
   ```
5. Upgrade to the `8.12.0` Agent.
   ```
   sudo elastic-agent upgrade 8.11.0 --source-uri file://$(pwd)/build/distributions/ --skip-verify
   ```
6. Ensure that the Agent version is now `8.12.0`.  Because the Agent will be crashing, you may only see the binary version, not the daemon version.
   ```
   sudo elastic-agent version
   ```
7. Tail the Upgrade Watcher logs. Ensure that the Upgrade Watcher eventually initiates rollback.
   ```
   sudo tail -F /opt/Elastic/Agent/data/elastic-agent-06be89/logs/elastic-agent-watcher-$(date +%Y%m%d).ndjson
   ```

8. Ensure that the rolled back Agent gets restarted.
9. Ensure that the Agent version is now back to `8.11.0`. 
   ```
   sudo elastic-agent version
   ```

## Related issues

- Fixes bug explained in https://github.com/elastic/elastic-agent/pull/3166#issuecomment-1661275131.
